src/realm/util/file.cpp

```diff
@@ -1293,6 +1305,34 @@ bool DirScanner::next(std::string& name)
     }
 }
 
+#elif REALM_HAVE_STD_FILESYSTEM
+
+DirScanner::DirScanner(const std::string& path, bool allow_missing)
+{
+    try {
+        m_iterator = std::filesystem::directory_iterator(path);
+    }
+    catch (const std::filesystem::filesystem_error& e) {
+        static std::error_condition no_such_directory(std::errc::no_such_file_or_directory);
+        if (e.code() != no_such_directory || !allow_missing)
+            throw;
+    }
+}
+
+DirScanner::~DirScanner() = default;
+
+bool DirScanner::next(std::string& name)
+{
+    static std::filesystem::directory_iterator end;
```

Comment:
`const`, and `static` probably isn't worth it (less likely to be inlined)

Reply:
I confess I have no idea what's the best way to go forward here - either compare directly with `directory_iterator()` or cache in a static variable. I went with the latter because that's what we (conservatively) do in C# to avoid GC churn, but if the conventional wisdom is to default-construct iterators for comparison purposes every time then I'm happy to do that instead.

Reply:
It's highly unlikely that the default constructor of the iterator is more expensive than a memory load from a static location, but in practice the compiler is probably clever enough that it doesn't make a difference. :-) Still, it might not be...

Reply:
This creates a race between simultaneous calls of DirScanner::next() ??

Reply:
I'll fix the static thing etc (it's also not thread safe)